### PR TITLE
[analyze] Analyze headers with the analyze skip file

### PR DESCRIPTION
**Pull request by csordasmarton**
Multiple source files may depend on a header file. If the skip file
given to the analyze command contains a header file with this patch
CodeChecker will try to find all the dependencies of these headers
and extend the skip file with these source files by analyzing these
source files too.

It can be used to analyze only the changed files properly. The only
limitation is that the full compilation database is needed to collect
these information.

### CODE REVIEW

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- analyzer/tests/functional/skip/test_skip.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5362094&bid=17570786)

File: analyzer/tests/functional/skip/test_skip.py [hunk 1]
```diff
@@ -111,3 +111,51 @@ def test_skip(self):
                                 "Found a location which points to "
                                 "skiped file, this report should "
                                 "have been removed.")
+
+    def test_analyze_only_header(self):
```
[the remark this answers is not in the document]
Comment (csordasmarton) [reply analyzer/tests/functional/skip/test_skip.py:108]: I created a separate test folder which will contain a header (lib.h) and 2 source files which depend on this header file and the test will check that both source files will be analyzed when the skip file includes only the header file.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- analyzer/tests/functional/skip/test_skip.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5362094&bid=17570786)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- analyzer/tests/functional/skip/test_skip.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5362094&bid=17570786)

File: analyzer/tests/functional/skip/test_files/simple/Makefile [hunk 2]
```diff
@@ -12,3 +12,5 @@ all: $(OBJS)
 
 clean:
 	rm -rf *.o
+
+multiple:
```
Comment (gyorb) [inline analyzer/tests/functional/skip/test_files/simple/Makefile:16]: Is this empty makefile target needed here?

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 3]
```diff
@@ -655,7 +675,47 @@ def __get_skip_handler(args):
...
+                norm_skip_path = os.path.normpath(skip_line[1:].strip())
+                rexpr = re.compile(
+                    fnmatch.translate(norm_skip_path + '*'))
+                analyze_headers.append((skip_line, rexpr))
+
+        # Get source files which depend on the previously collected header
+        # files and create a new skip list handler where we include these
+        # files in the beginning.
+        if analyze_headers:
+            dependencies = __get_source_dependencies(compile_commands)
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/cmd/analyze.py:697]: I think an info message would nice here so the user will know what is happening. It takes a few seconds without any output message to collect the dependencies.

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 4]
```diff
@@ -655,7 +675,47 @@ def __get_skip_handler(args):
...
+            for f, deps in dependencies.items():
+                for _, rexpr in analyze_headers:
+                    if rexpr.match(f):
+                        analyze_header_deps.extend(["+" + d for d in deps])
+
+            if analyze_header_deps:
+                LOG.info("Your skip file contained some header files (%s). "
+                         "For this reason CodeChecker will include the "
+                         "following source files which depends on these "
+                         "header files:\n%s",
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/cmd/analyze.py:709]:
```suggestion
                LOG.info("Your skip file contained some header files (%s) to be analyzed. "
                         "Analysis can not be executed on header files only. "
                         "For this reason CodeChecker will analyze the "
                         "following source files which include the header files "
                         ":\n%s",
```
I thinks a bit more explanation to the user would be nice here to better understand why those source files will be analyzed.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- analyzer/tests/functional/skip/test_skip.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5362094&bid=17570786)

**Pull request merged**